feat(op-acceptance-tests): add more sysgo tests.

scharissis opened the pull request:
Adds a lot more sysgo tests (all of them)!
This is as part of the effort to unify all acceptance tests into the one workflow in CI.

<img width="833" height="475" alt="Screenshot 2025-08-22 at 09 09 15" src="https://github.com/user-attachments/assets/9594e356-b938-4424-8ad6-c8d28190a960" />

scharissis committed:
feat(op-acceptance-tests): add more sysgo tests.

opgitgovernance commented:
This pr has been automatically marked as stale and will be closed in 5 days if no updates

scharissis committed:
fix(op-devstack): ecotone; fees.

Use the actual increase in the L1FeeVault balance. This is because in Ecotone, the L1 fee includes both base fee and blob base fee components.

scharissis committed:
fix(op-acceptance-tests): skipping flaky/broken tests.

scharissis committed:
fix(op-acceptance-tests): interop; TestInteropSystemSupervisor

scharissis committed:
fix(op-acceptance-tests): justfile; exit early for gateless.

scharissis committed:
fix(op-acceptance-tests): tidied justfile

scharissis committed:
fix(op-acceptance-tests): fix ecotone fees.

The recent commit changed L1 fee calculation to use vault balance increases as the source of truth, but base fee and priority fee calculations were still using the old method (calculating from block data). Made all fee calculations by making vault increases the source of truth.

scharissis committed:
fix(op-acceptance-tests): increase default timeout for gateless mode.

janjakubnanista committed:
chore: Conditional logic

scharissis committed:
fix(op-acceptance-tests): TestSuperRootWithdrawal nonce issue and timeouts

pcw109550 commented on op-acceptance-tests/tests/safeheaddb_elsync/safeheaddb_test.go:
```diff
@@ -10,6 +10,8 @@ import (
 )
 
 func TestTruncateDatabaseOnELResync(gt *testing.T) {
+	gt.Skip("TODO(#17083): skipping until we have a fix")
```
https://github.com/ethereum-optimism/optimism/issues/17083 is resolved.

scharissis replied:
Thanks! Re-enabled in [11aaf6d](https://github.com/ethereum-optimism/optimism/pull/16817/commits/11aaf6d50a976a4b6c6892d1a13575d4a26b77a9) 

Unfortunately it failed again: https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/98811/workflows/69d410ac-9c8d-45e1-b949-223d271d83f5/jobs/3799311/artifacts

```
...
    assertions.go:369:             [31mERROR[0m[08-21|23:37:21.352] 
    assertions.go:369:             	Error Trace:	/home/circleci/project/op-devstack/dsl/safedb.go:22
    assertions.go:369:             	            				/home/circleci/project/op-devstack/dsl/l2_cl.go:335
    assertions.go:369:             	            				/home/circleci/project/op-acceptance-tests/tests/safeheaddb_elsync/safeheaddb_test.go:51
    assertions.go:369:             	Error:      	Expected value not to be nil.
    assertions.go:369:             	Test:       	TestNotTruncateDatabaseOnRestartWithExistingDatabase
    assertions.go:369:             	Messages:   	no safe head data available at L1 block 5
    assertions.go:369:             
--- FAIL: TestNotTruncateDatabaseOnRestartWithExistingDatabase (22.18s)
```

Should we address this in a future PR perhaps?

scharissis replied:
Thanks! Re-enabled in [11aaf6d](https://github.com/ethereum-optimism/optimism/pull/16817/commits/11aaf6d50a976a4b6c6892d1a13575d4a26b77a9)

pcw109550 commented on op-acceptance-tests/tests/safeheaddb_elsync/safeheaddb_test.go:
```diff
@@ -39,6 +41,8 @@ func TestTruncateDatabaseOnELResync(gt *testing.T) {
 }
 
 func TestNotTruncateDatabaseOnRestartWithExistingDatabase(gt *testing.T) {
+	gt.Skip("TODO(#17083): skipping until we have a fix")
```
ditto

scharissis replied:
Thanks! Re-enabled in [11aaf6d](https://github.com/ethereum-optimism/optimism/pull/16817/commits/11aaf6d50a976a4b6c6892d1a13575d4a26b77a9) 

Unfortunately it failed again: https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/98811/workflows/69d410ac-9c8d-45e1-b949-223d271d83f5/jobs/3799311/artifacts

```
...
    assertions.go:369:             [31mERROR[0m[08-21|23:37:21.352] 
    assertions.go:369:             	Error Trace:	/home/circleci/project/op-devstack/dsl/safedb.go:22
    assertions.go:369:             	            				/home/circleci/project/op-devstack/dsl/l2_cl.go:335
    assertions.go:369:             	            				/home/circleci/project/op-acceptance-tests/tests/safeheaddb_elsync/safeheaddb_test.go:51
    assertions.go:369:             	Error:      	Expected value not to be nil.
    assertions.go:369:             	Test:       	TestNotTruncateDatabaseOnRestartWithExistingDatabase
    assertions.go:369:             	Messages:   	no safe head data available at L1 block 5
    assertions.go:369:             
--- FAIL: TestNotTruncateDatabaseOnRestartWithExistingDatabase (22.18s)
```

Should we address this in a future PR perhaps?

scharissis replied:
Thanks! Re-enabled in [11aaf6d](https://github.com/ethereum-optimism/optimism/pull/16817/commits/11aaf6d50a976a4b6c6892d1a13575d4a26b77a9) 

Unfortunately it failed again: https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/98811/workflows/69d410ac-9c8d-45e1-b949-223d271d83f5/jobs/3799311/artifacts

```
...
    assertions.go:369:             [31mERROR[0m[08-21|23:37:21.352] 
    assertions.go:369:             	Error Trace:	/home/circleci/project/op-devstack/dsl/safedb.go:22
    assertions.go:369:             	            				/home/circleci/project/op-devstack/dsl/l2_cl.go:335
    assertions.go:369:             	            				/home/circleci/project/op-acceptance-tests/tests/safeheaddb_elsync/safeheaddb_test.go:51
    assertions.go:369:             	Error:      	Expected value not to be nil.
    assertions.go:369:             	Test:       	TestNotTruncateDatabaseOnRestartWithExistingDatabase
    assertions.go:369:             	Messages:   	no safe head data available at L1 block 5
    assertions.go:369:             
--- FAIL: TestNotTruncateDatabaseOnRestartWithExistingDatabase (22.18s)
```

Should we address this in a future PR perhaps?

scharissis replied:
This will hopefully be fixed by https://github.com/ethereum-optimism/optimism/pull/17179

scharissis replied:
Seems to be fixed now :)

pcw109550 commented on op-acceptance-tests/justfile:
```diff
@@ -19,27 +21,27 @@ interop:
+    GATELESS_MODE=false
+    if [[ "{{gate}}" == "" ]]; then
+        GATELESS_MODE=true
+        echo -e "DEVNET: in-memory (sysgo), MODE: gateless (all tests)\n"
```
For `just acceptance-test simple ""`, my understanding for this command is external devnet(sysext) + gateless(all tests) but this conditionals fall to sysgo. Maybe introduce a double nested branch?

scharissis replied:
Added support for running gateless mode with both sysext and sysgo in [96b2b6a](https://github.com/ethereum-optimism/optimism/pull/16817/commits/96b2b6a8d2296f07b3bd03fb9e36a02c0c1f2780)

scharissis replied:
Resolving to merge. But let me know if there's any outstanding issues!

pcw109550 commented on .circleci/config.yml:
```diff
@@ -2522,6 +2527,17 @@ workflows:
+      # IN-MEMORY (all)
+      - op-acceptance-tests:
+          name: memory-all
+          gate: ""  # Empty gate = gateless mode
```
Should we remove the memory-base flow since memory-all may contain memory-base?

scharissis replied:
Yes, removed.

pcw109550 commented on op-acceptance-tests/tests/interop/proofs/withdrawal/withdrawal_test.go:
```diff
@@ -29,6 +29,9 @@ func TestSuperRootWithdrawal(gt *testing.T) {
 	l1User.VerifyBalanceExact(initialL1Balance.Sub(depositAmount).Sub(deposit.GasCost()))
 	l2User.VerifyBalanceExact(initialL2Balance.Add(depositAmount))
 
+	// Wait for a block to ensure nonce synchronization between L1 and L2 EOA instances
```
Can you please elaborate on `nonce synchronization`? It would be helpful if what kind of flakiness was experienced.

scharissis replied:
I don't recall much, sorry. Essentially what I observed was a lot of flakiness (maybe even 100% failure rate actually) with a few tests, such as this one, in CI. But locally they worked fine.
I tried adding a few synchronizations, like above, plus increasing a few timeouts and that seemed to help.
The log files were huge so it was hard for me to determine the specific root cause. Nothing stood out in particular.

I'd be more than happy to re-create this and investigate with you.

pcw109550 replied:
Left comment at https://github.com/ethereum-optimism/optimism/pull/16817#discussion_r2297611747

pcw109550 commented on op-devstack/dsl/ecotone_fees.go:
```diff
@@ -58,24 +58,27 @@ func (ef *EcotoneFees) ValidateTransaction(from *EOA, to *EOA, amount *big.Int)
-		l1Fee = receipt.L1Fee
-	}
+	// In Ecotone, L1 fee includes both base fee and blob base fee components
+	l1Fee := vaultIncreases.L1FeeVault // Use actual vault increase as the source of truth
```
Can we still check that the receipt included L1 fee is equal to the `vaultIncreases.L1FeeVault`?

scharissis replied:
Hopefully addressed in [e0feddc](https://github.com/ethereum-optimism/optimism/pull/16817/commits/e0feddc48d33b660574f9e8bf6a96b4cf35c91b3). Please take a look and let me know if this addresses your feedback.

pcw109550 replied:
Left comment at https://github.com/ethereum-optimism/optimism/pull/16817#discussion_r2297602747

pcw109550 commented on op-devstack/dsl/ecotone_fees.go:
```diff
@@ -58,24 +58,27 @@ func (ef *EcotoneFees) ValidateTransaction(from *EOA, to *EOA, amount *big.Int)
+	l2Fee := new(big.Int).Add(baseFee, priorityFee)
 
-	baseFee := new(big.Int).Mul(block.BaseFee(), big.NewInt(int64(receipt.GasUsed)))
-	l2Fee := new(big.Int).Mul(receipt.EffectiveGasPrice, big.NewInt(int64(receipt.GasUsed)))
```
Also this, we need to preserve the checks for the receipt included fee data like `receipt.EffectiveGasPrice`.

scharissis replied:
Hopefully addressed in [e0feddc](https://github.com/ethereum-optimism/optimism/pull/16817/commits/e0feddc48d33b660574f9e8bf6a96b4cf35c91b3). Please take a look and let me know if this addresses your feedback.

pcw109550 replied:
My idea was comparing baseFee, receiptL2Fee  to receiptBaseFee, l2Fee. so
assert baseFee == receiptBaseFee and receiptL2Fee == l2Fee.
Does it make sense to you? So we may sanity check the fee vaults and the receipt fields by this check.

pcw109550 commented on op-devstack/dsl/ecotone_fees.go:
```diff
@@ -58,24 +58,27 @@ func (ef *EcotoneFees) ValidateTransaction(from *EOA, to *EOA, amount *big.Int)
+	// Total fee is the sum of all vault increases
+	totalFee := new(big.Int).Add(vaultIncreases.BaseFeeVault, vaultIncreases.L1FeeVault)
+	totalFee.Add(totalFee, vaultIncreases.SequencerVault)
+	totalFee.Add(totalFee, vaultIncreases.OperatorVault)
```
My understanding is that the operator fee was added at isthmus, after ecotone. So do not need the `OperatorVault` check at all. 
Ref) https://specs.optimism.io/protocol/isthmus/predeploys.html#operatorfeevault

We should check the increment `vaultIncreases.OperatorVault` must be zero at ecotone.

scharissis replied:
Hopefully addressed in [e0feddc](https://github.com/ethereum-optimism/optimism/pull/16817/commits/e0feddc48d33b660574f9e8bf6a96b4cf35c91b3). Please take a look and let me know if this addresses your feedback.

pcw109550 replied:
Left comment at https://github.com/ethereum-optimism/optimism/pull/16817#discussion_r2297602747

pcw109550 commented on op-devstack/dsl/ecotone_fees.go:
```diff
@@ -58,24 +58,27 @@ func (ef *EcotoneFees) ValidateTransaction(from *EOA, to *EOA, amount *big.Int)
 	vaultsAfter := ef.getVaultBalances(client)
 	vaultIncreases := ef.calculateVaultIncreases(vaultsBefore, vaultsAfter)
 
-	l1Fee := big.NewInt(0)
```
Overall comment for the `ValidateTransaction` method: There is a update for fee check. May I ask what kind of flakiness did you experience? Asking this because curious that this patch is towards reducing flakiness, or hardening test, or both.

scharissis replied:
(copied from a previous response above, as the same applies)

I don't recall much, sorry. Essentially what I observed was a lot of flakiness (maybe even 100% failure rate actually) with a few tests, such as this one, in CI. But locally they worked fine.
I tried adding a few synchronizations, like above, plus increasing a few timeouts and that seemed to help.
The log files were huge so it was hard for me to determine the specific root cause. Nothing stood out in particular.

I'd be more than happy to re-create this and investigate with you.

pcw109550 replied:
Sharing my ideas.

> The log files were huge so it was hard for me to determine the specific root cause

This seems like an issue we should address. From what I understand, we used to run tests with **gotestsum**, which presented logs in a clearer way. With **op-acceptor**, log-based debugging seems more difficult. Could you share some context on why the switch was made, and whether we have options to improve log readability under the new setup?

> I tried adding a few synchronizations, like above, plus increasing a few timeouts and that seemed to help.

Do we have a working hypothesis about the underlying cause (for example race condition)? Otherwise, we risk the test flaking again after merge. A concrete explanation, or an adjustment that waits on a deterministic condition rather than longer timeouts would give more confidence in the fix.

scharissis replied:
>  This seems like an issue we should address. From what I understand, we used to run tests with gotestsum, which presented logs in a clearer way. With op-acceptor, log-based debugging seems more difficult. Could you share some context on why the switch was made, and whether we have options to improve log readability under the new setup?

This is interesting to hear! My understanding, and experience, is that when we run with `gotestsum` the logs are harder to debug with.
But this change was motivated much more by the principle of having them all in one place and unifying the logging. If op-acceptor-based debugging is more difficult then we should address that! If now, or in the future, you have any concrete examples/issues here please forward them to me.
We can almost certainly improve the readability if we know what the issue(s) is/are.

> Do we have a working hypothesis about the underlying cause (for example race condition)? Otherwise, we risk the test flaking again after merge. A concrete explanation, or an adjustment that waits on a deterministic condition rather than longer timeouts would give more confidence in the fix.

I do not. But I did notice that they were a LOT less flaky (or even fixed) after increasing the timeouts.
I am happy to help pursue the discovery of a root cause here, but I am not well equipped to do it myself at this time.

zhiqiangxu commented on op-acceptance-tests/justfile:
```diff
@@ -63,6 +65,18 @@ acceptance-test devnet="" gate="holocene":
+        echo "mise is installed"
+    else
+        echo "Mise not installed, falling back to Docker..."
+        just acceptance-test-docker {{devnet}} {{gate}}
```
Looks like the `acceptance-test-docker` recipe doesn't support the `"{{devnet}}" == ""` case ?

scharissis replied:
You're correct. Thank you for bringing it to my attention!

scharissis replied:
I'll address this in a future PR because the docker fallback is generally very stale and needs a lot of love.
I'm not actually sure it's even used.

scharissis committed:
fix(op-acceptance-tests): re-enable safeheaddb tests after #17083 fix

scharissis committed:
fix(op-acceptance-tests): security; env var injection

The LOG_LEVEL environment variable is used directly in shell command execution without validation or sanitization, allowing command injection attacks.

scharissis committed:
feat(op-acceptance-tests): justfile; allow gateless for sysext.

Allow gateless mode to run for an external devnet ('sysext' orchestrator)

scharissis committed:
clean(op-acceptance-tests): circleci; removed memory-base job

The new memory-all is a superset of memory-base; so we don't need both.

scharissis committed:
fix(op-devstack): enhance Ecotone fee validation and enforce operator vault constraints

  - Add validation that receipt L1Fee matches L1FeeVault increase
  - Restore receipt-based fee calculations for validation (baseFee and L2Fee)
  - Enforce OperatorVault must be zero in Ecotone (operator fees introduced in Isthmus)
  - Exclude OperatorVault from total fee calculations in Ecotone
  - Add comprehensive receipt fee validation checks

scharissis committed:
fix(op-devstack): ecotone; cross-validation of fees

Added cross-validation between receipt and vault fees: The code now verifies that:
    - receiptBaseFee (block.BaseFee × gasUsed) equals vaultBaseFee (BaseFeeVault increase)
    - receiptL2Fee (effectiveGasPrice × gasUsed) equals vaultL2Fee (BaseFee + SequencerFee vault increases)

scharissis committed:
fix(op-devstack): supervisor sync status L1 mismatch retry

Increase retry attempts and handle L1 sync mismatch errors to fix flaky TestExecMsgDifferEventIndexInSingleTx.

scharissis commented on op-acceptance-tests/justfile:
```diff
@@ -74,20 +89,17 @@ acceptance-test devnet="" gate="holocene":
+            --orchestrator sysgo \
+            --allow-skips \
+            --timeout 90m \
+            --log.level "${LOG_LEVEL:-info}"
```
This has been adopted, just slightly differently.

pcw109550 reviewed (approved):
Approving with the current CI experience.

https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/99120/workflows/df8e73b4-afc9-4e6b-88de-5fdc67213e99 is the latest re-run for the `acceptance-tests-pr` -> `memory-all` pipeline.

The CI is red.
<img width="915" height="260" alt="image" src="https://github.com/user-attachments/assets/52e74645-3788-4df0-b74c-323207264b0b" />

when I click the `memory-all` pipeline,
<img width="1471" height="210" alt="image" src="https://github.com/user-attachments/assets/96419c42-b89c-405f-88a8-4f78a06f1901" />

It shows `Exited with code exit status 1`

Also the test tab does not show the failed test:
<img width="845" height="360" alt="image" src="https://github.com/user-attachments/assets/fc7d274c-a6b5-4321-b8bf-34f931be89ea" />

I can view the artifacts tab:
<img width="473" height="485" alt="image" src="https://github.com/user-attachments/assets/a5141962-bd93-4929-91f4-3ff78c4c65b1" />

And directly view which tests failed: `TestChallengerPlaysGame`. I can also inspect the logs for it, by clicking the log file link.

So this PR improves the log granularity of CI, improving debugging experience 👍, and also unifying the method of running acceptance tests using op-acceptor. One suggestion it may be nice that the CircleCI frontend itself shows the failed test case(at the test tab), but its really good to have. Nice work!

scharissis committed:
fix(op-acceptance-tests): TestPostInteropUpgradeComprehensive

Made it serial due to flakiness.

scharissis committed:
feat(op-acceptance-tests): ci; tweak timeout and verbosity.

scharissis committed:
fix(op-acceptance-tests): exclude interop tests from ci tag

These in-memory devstack tests don't work reliably with gotestsum's
retry mechanism. They run fine via op-acceptor in the memory-all job
but fail when run directly with go test in go-tests-short.

Adding !ci build constraint to exclude from standard CI test suite
while keeping them available for dedicated acceptance test runs.

scharissis committed:
fix(op-acceptance-tests): TestWithdrawal nonce synchronization

Use fresh EOA instance for withdrawal initiation to prevent nonce
conflicts when shared L1/L2 keys encounter retry logic.

scharissis committed:
chore(op-acceptance-tests): TestPreNoInbox marked as flaky

scharissis committed:
chore(op-acceptance-tests): test limiting concurrency

scharissis committed:
chore(op-acceptance-tests): self-hosted runner

Default to using our self-hosted runners.
Accept a parameter which allows individual jobs to opt-out and use CCI/Cloud runners.

nonsense commented:
This is great and I love that we are unifying the CI, but the `memory-all` job is taking on occasion > 45min. - for example https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/100022/workflows/901df824-9c6d-421c-94b9-f325452489b4/jobs/3837984

We should work to get to a more reasonable CI execution time, I think not more than 10-15min.

scharissis commented:
> This is great and I love that we are unifying the CI, but the `memory-all` job is taking on occasion > 45min. - for example https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/100022/workflows/901df824-9c6d-421c-94b9-f325452489b4/jobs/3837984
> 
> We should work to get to a more reasonable CI execution time, I think not more than 10-15min.

Absolutely; agreed!

I noticed this before merging but chose to merge anyway as I believe it's a net improvement.
As you note, this job doesn't _always_ last this long (eg. here's a [17min run](https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/100139/workflows/ea3b379a-0551-466a-a24d-3885b31e8633)), so I suspect there's some issues with the tests whereby one or more almost deadlocks.

I'll work to improve our observability here (eg. the logging, as you noted elsewhere) so we can identify and remedy this.
